### PR TITLE
add `generateName` to devfile metadata

### PR DESCRIPTION
**Pull request by sparkoo**
### What does this PR do?
Adds `generateName` field to devfile. The field is used just in create workspace phase, where it's transformed into regular devfile `name`. The `generateName` is not persisted nor used in later phases like workspace startup, update, etc.
Current implementation follows Kubernetes with `[generateName]YYYYY`, where `Y` is random `[a-z0-9]` character.

### What issues does this PR fix or reference?
#13794 (see comment https://github.com/eclipse/che/issues/13794#issuecomment-518549557)

#### Docs PR
https://github.com/redhat-developer/devfile/pull/46
https://github.com/redhat-developer/devfile/pull/48

### CODE REVIEW

File: wsmaster/che-core-api-workspace/src/main/resources/schema/devfile.json [hunk 1]
```diff
@@ -50,10 +50,26 @@
           "examples": [
             "petclinic-dev-environment"
           ]
+        },
+        "generateName": {
+          "type": "string",
+          "minLength": 1,
+          "title": "Devfile Generate Name",
+          "description": "Workspaces created from devfile, will have use it as base and append random suffix. It's used when name is not defined.",
```
Comment (benoitf) [inline wsmaster/che-core-api-workspace/src/main/resources/schema/devfile.json]:
```suggestion
          "description": "Workspaces created from devfile, will use it as base and append random suffix. It's used when name is not defined.",
```

File: wsmaster/che-core-api-workspace/src/test/resources/devfile/devfile.yaml [hunk 2]
```diff
@@ -14,6 +14,7 @@
 apiVersion: 1.0.0
 metadata:
   name: petclinic-dev-environment
```
Comment (benoitf) [inline wsmaster/che-core-api-workspace/src/test/resources/devfile/devfile.yaml:16]: is it valid to have both name and generateName ?

Comment (sparkoo) [reply wsmaster/che-core-api-workspace/src/test/resources/devfile/devfile.yaml:16]: it is valid, but 'generateName' is basically ignored in that case.

**Review by skabashnyuk — COMMENTED** [review]
Looks good to  me

**Comment by sparkoo** [comment]
ci-test

**Comment by dmytro-ndp** [comment]
@sparkoo: PR check has been failed because it uses Jenkinsfile now which has been merged in master tonight. So, you need to take changes from master.

**Comment by sparkoo** [comment]
ci-test

**Comment by sleshchenko** [comment]
@sparkoo Added WIP in PR title to avoid occasional merge

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java [hunk 3]
```diff
@@ -569,6 +575,40 @@ private WorkspaceImpl doCreateWorkspace(
...
+    }
+    return origDevfile;
+  }
+
+  /**
+   * If 'generateName' is defined and 'name' is not, we generate name using 'generateName' as a
+   * prefix following {@link Constants#WORKSPACE_GENERATE_NAME_CHARS_APPEND} random characters and
+   * set it to 'name'.
+   */
+  private DevfileImpl generateNameIfNeeded(DevfileImpl devfileModel) {
```
Comment (sleshchenko) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java]: Seems it's unused.

**Review by sleshchenko — APPROVED** [review]
LGTM

Please take a look my inlined comments

**Comment by skabashnyuk** [comment]
ping @che-bot

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java [hunk 4]
```diff
@@ -569,6 +575,40 @@ private WorkspaceImpl doCreateWorkspace(
...
+      if (metadata.getName() == null && metadata.getGenerateName() != null) {
+        metadata.setName(
+            NameGenerator.generate(
+                metadata.getGenerateName(), WORKSPACE_GENERATE_NAME_CHARS_APPEND));
+        DevfileImpl devfileWithGeneratedName = new DevfileImpl(origDevfile);
+        devfileWithGeneratedName.setMetadata(metadata);
+        return devfileWithGeneratedName;
+      }
+    }
+    return origDevfile;
```
[the remark this answers is not in the document]
Comment (sparkoo) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java:595]: It is not valid, but it is not responsibility of this method to validate devfile. Validation happens before.

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java (hunk 3 above)
Comment (sparkoo) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java]: done

File: core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/workspace/devfile/Metadata.java [hunk 5]
```diff
@@ -16,4 +16,6 @@
 
   /** @return the name of the devfile */
   String getName();
+
```
[the remark this answers is not in the document]
Comment (sparkoo) [reply core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/workspace/devfile/Metadata.java:19]: done

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java [hunk 6]
```diff
@@ -569,6 +575,26 @@ private WorkspaceImpl doCreateWorkspace(
     return workspace;
   }
 
+  /**
+   * If 'generateName' is defined and 'name' is not, we generate name using 'generateName' as a
```
Comment (sunix) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java:579]: Could we use `generateName` all the time if it is defined ?

Comment (sparkoo) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java:579]: kubernetes takes `name` in that case and we follow kubernetes approach here.

**Review by benoitf — APPROVED** [review]

**Pull request merged**